From 0fe6324515531410b291c6734434cd6b89e77155 Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Mon, 9 Aug 2021 15:24:35 +0000
Subject: [PATCH backport v.4.19 2/4] mlxsw: minimal: Obtain maximum number of
 slots supported by system

Obtain maximum slot number by query MGPIR register with 'slot_index'
zero. This info is to be used to pre-allocate memory for maximum line
card number during initialization.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/net/ethernet/mellanox/mlxsw/minimal.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/minimal.c b/drivers/net/ethernet/mellanox/mlxsw/minimal.c
index e5b1bc90b..599ccf89d 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/minimal.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/minimal.c
@@ -253,6 +253,14 @@ static int mlxsw_m_ports_create(struct mlxsw_m_area *mlxsw_m_area, u8 slot_index
 	if (err)
 		return err;
 
+	/* Get maximum number of slots supported by system. */
+	if (!slot_index)
+		mlxsw_m->max_modules_per_slot =
+			mlxsw_reg_mgpir_max_modules_per_slot_get(mgpir_pl);
+	else
+		mlxsw_m_area->module_off = (slot_index - 1) *
+					   mlxsw_m->max_modules_per_slot;
+
 	mlxsw_reg_mgpir_unpack(mgpir_pl, NULL, NULL, NULL,
 			       &mlxsw_m_area->max_ports, NULL);
 	if (!mlxsw_m_area->max_ports)
-- 
2.20.1

